design/gateway-apis-implementation.md

```diff
@@ -1,204 +1,420 @@
+Doing so will leave the fleet as-is and stop taking action against the Envoy fleet.
+
+#### managed: true
+
+If envoy.managed is true, then Contour will automatically create an Envoy fleet/service.
+
+If Gateway.ControllerName is not set, then the envoy-service-xxx flags passed to Contour will be used to define the ports which Envoy/Service will utilize.
+
+If `Gateway.ControllerName` is set then the ports that are assigned to the Envoy service will match the Gateway discovered based upon the ControllerName.
+
+Additional RBAC permissions will be required to the namespace where Contour is deployed to allow it to dynamically manage Daemonsets/Deployments/Services, etc.
+
+#### managed: false
+
+If envoy.managed is false, then Contour won't manage the Envoy fleet/service. It's up to the cluster operator to deploy and manage these resources.
+The external ports of the load balancer must match the ports Contour expects (otherwise insecure-secure redirects won’t work).
 
-### Code changes
-Contour already has support for importing the Gateway API objects into its Kubernetes cache.
-However, for some types, we need to be able to keep some details - this design suggests making those details properties of the cache, as `IngressClass` is currently.
+The external ports should almost always be 80/443.
 
-For ingestion, the general pattern currently is that the EventHandler in `internal/contour/handler.go` handles all objects, and calls out to the KubernetesCache from `internal/dag/cache.go`.
-The current pattern across most of the objects is:
-- check if the object is in scope
-- add it to the cache
-- indicate with a return value if the add should result in a DAG (and consequently an Envoy config) rebuild.
+If Gateway.ControllerName is not set, then the envoy-service-xxx flags passed to Contour will be used to define the ports which Envoy/Service will utilize.
 
-This pattern is applicable to all the Gateway API objects, with the exception of GatewayClass, which is not relevant for Contour.
+See: [https://github.com/projectcontour/contour/issues/3616](https://github.com/projectcontour/contour/issues/3616)
 
-There are already internal fields in the cache to hold GatewayClass and some other Gateway API objects.
-GatewayClass will be removed as part of this work, then we will use this scaffolding to build the actual functionality on top.
+### Leader Election
+
+Contour will take into account if it is the Leader or not which will represent if it should take action on managing resources or not.
+
+### Finalizers
+
+Finailzers will need to be put into place for the resources which Contour owns (i.e. Envoy) as well a GatewayAPI bits if applicable:
+
+* Envoy Service/Daemonset
+* Gateway
+* GatewayClass
+
+#### Finalizer workflows:
+
+* If the Envoy service is deleted what happens?
+* If the Envoy Daemonset is deleted what happens?
+* If the Gateway is deleted what happens?
+* If the GatewayClass is deleted what happens?
+
+### Exposing Envoy
```

Comment:
had a thought about this and what the Gateway listener ports "mean" and how we could service multiple scenarios of deployments

how much do users *need to customize* the ports Envoy listens on in the container/how much do we have to think about how the Envoy pods are routable?

If we choose a certain direction, are we locking people who use a managed Envoy to do a specific thing, which may be incompatible with how they have their cluster set up (routable pods or not, CNI, other constraints)?

if the ports on a Gateway listener represent the ports that a user outside a cluster will access apps via and we choose Envoy internal ports for you/make you configure them outside of the Gateway, that seems unwieldy potentially?

also, if we change how we treat those ports (are they on the LB service or just what envoy listens on) halfway through implementation/migration that will probably not be a great experience

thinking about this a bit and seeing all the places we have port configurations etc. in our example deployments (that people right now have the ability to customize to make work) makes me think we need to start earlier with the "gateways everywhere" approach for managed Envoy, and reconcile 2 different types of Gateways, one for configuring how Envoy listens, and another for how Envoy is exposed
- a migration step here could be to keep the Envoy LoadBalancer service and associated implementation logic a bit separate from the Gateway stuff, and make that for users using separate configuration
- another approach would be to just include it in example YAMLs and leave it up to the user how they want traffic into Envoy

Reply:
this was discussed a bit in the doc, but thought I would bring it up again, I dont know if we fully discussed the various migration steps to the end state of things and how all these ports interact when youre not using Gateway API though

Reply:
just worried we get locked into one deployment model or cluster configuration and find it hard to adapt later b/c theres no space in the API for adding more information to a Gateway for example

Reply:
I think it's not stated very clearly, but the assumption is that the Gateway's ports are the ports that the traffic should be sent to from the outside of the Gateway, and that doing other stuff here is implementation detail. I don't think that having the Gateway ports represent the actual container ports Envoy is running on is a good idea without a really good reason.

I think that for this bit, we should start with the easy case, and build outward.

The ports declared on Contour's Gateway are the externally-visible ports that traffic should be sent to via a client. When we are managing the Envoy inbound traffic using a Service or second Gateway object, those should be the requested ports, and we should use a port-translation mechanism like `targetPort` in Service to translate to a high port on the Envoy deployment. As we add more complicated Service requests in, we need to keep this in mind, that the *outer* ports of the Service need to be the same as the ones on the Contour Gateway.

People can customise those ports right now in our example YAMLs, and if they want to do that, managed Envoy is not ready for them yet. I think this is okay to start with. Then, people can tell us what they need, and we can help make that work.

The fact we have to build that is an unavoidable cost of adding the Envoy management functionality in to Contour, sadly.

It's really important to get this right because otherwise we'll break the insecure->secure redirects as well.

Reply:
tl;dr I think we should go ahead with just LB Service to start with, and consider the container ports and mappings an implementation detail.

Reply:
Feels like a missing bit in the gateway design since an ingress controller doesn't really manage the ports, the external gateway does. 

An ingress gateway us just a cluster service at this point.

Reply:
I'll update the design. Can start with type loabalancer, but Im going to add a spot to reference a gateway instead. We need to do that, seems silly to not enable gateways when implementing gateways. 🙃